### ISSUE TITLE
The implementation is quite different from the paper?

### ISSUE THREAD

**Issue by BluebirdStory** [qianyuzqy/IADG#12]
About AIAW in the paper. In paper, it's said that the feature will first go through a Instance Norm layer to calculate covariance matrix, and then, the std of covariance matrix will be used as a clue for style-sensitive feature cov finding.
However, is actual implementation, it's only just a batchNorm and a torch.bmm for gram matrix computing.

I wonder why there is a such huge difference?

thank you!

**Comment by qianyuzqy** [comment]
HI, @BluebirdStory ,
I am very sorry for your misunderstanding of our method. 
1) As for the Instance Norm layer, it is in the L51 of base_block.py. The feature map is fed into an instance normalization (IN) layer to output a normalized feature, and the concat of the normalized feature (L130 of framework.py)) is used to calculate feature correlation.
2) As for the feature covariance, we follow the same implementation as [A,B] (i.e., gram matrix) for calculating the feature correlation.  In addition, the role of batchNorm in self.conv_final is to project the original feature and augmented feature to the same space for computing feature covariance.
[A]. RobustNet: Improving Domain Generalization in Urban-Scene Segmentation via Instance Selective Whitening, CVPR 2021;
[B].https://github.com/shachoi/RobustNet/blob/1ff5c047d7abaf9ab1a7cca23aff812b2f09de15/network/instance_whitening.py#L30

**Comment by BluebirdStory** [comment]
Thank you for your patience and excellent work!
I have throughly read RobustNet, your paper and implementation. I still have the following questions.
1. I have read base_block.py, for your FeatExtractor, it's a pure-instanceNorm CNN. However, after your IN, it's always followed by a ReLU act, which will eliminate value below zero. Besides, after self.conv_final, which consists of a Conv3x3 and BN, is it still safe to save a covariance matrix is calculated in Formula(6)? I think it's actual Gram Matrix.
2. In Formula(7), the paper said each sample will be iterated to calculate the correlation std, in my understanding, this calculation makes sense only when a feature and its exact style-augmented version is involved(only consider the style, not content), however, in your implementation, only a torch.std on Gram(BxCxC) is used, both style and content is involved, I wonder if it is reasonable ?
3. the last but not least, in your implementation of cov mask generation(cov_settings.py L69), it seems that every epoch's cov mask will consider the past epochs' counterpart, I wonder why there is such design.
code: self.mask_matrix = (self.mask_matrix.int() & mask_matrix.view(self.dim, self.dim).int()).float()

thank you

**Comment by BluebirdStory** [comment]
@qianyuzqy hate to interrupt ..., but any further info?

**Comment by qianyuzqy** [comment]
@BluebirdStory I'm sorry that I've been sick recently. I've been in the hospital for the last few days, and I am out of the office. I will reply to your message in about one week. Thank you for your understanding.

**Comment by BluebirdStory** [comment]
@qianyuzqy really sorry to hear that..., wish you will be all well soon. no rush, I can wait until you are all well. anyway, thank you for your patience and excellent work, wish you good health!

**Comment by BluebirdStory** [comment]
Sorry to bother, how is your health? is everything allright?

**Comment by BluebirdStory** [comment]
@qianyuzqy hi bro, hate to bother, is everything all right?

**Comment by qianyuzqy** [comment]
@BluebirdStory Sorry for the late reply. I would like to answer your questions as follows:
1) As mentioned in Line 19, Page 3 of RobustNet, "a gram matrix or covariance matrix is feature correlation that takes style information of images" and we follow the same implementation as ReboustNet [A] in implementing the feature correlation. In Page 5, the authors refer to this matrix as a covariance matrix. We acknowledge you pointing out this issue, we admit that using the name of feature correlation is better and more suitable.

2)  Sorry for the confusion. We have already augmented the styles via the CSA module and thus it is reasonable to calculate the feature correlation between the feature and its exact style-augmented version. Besides, we follow a similar manner of computing instance whitening as RobustNet.

3) Sorry for the confusion.  This design is also borrowed from RobustNet [A]. In fact, we don't consider the previous cov mask during the training. Since the feature space is dynamically changing along with the model updating and thus the cov mask needs to be updated every epoch after the style bases are chosen.

Thank you very much for your interest in our paper！

Best Regards,
